### PR TITLE
feat: add error details to rest and api exceptions

### PR DESCRIPTION
**Pull request by ashish-s**
# Feature #
Add a new property named details to RestException add ApiException. Details is returned by selective apis to populate extra information about errors.
The Studio V2 flow api will be returning extra details with flow validation errors.
### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guidelines](CONTRIBUTING.md) and my PR follows them
- [x] I have titled the PR appropriately
- [x] I have updated my branch with the master branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation about the functionality in the appropriate .md file
- [x] I have added inline documentation to the code I modified

If you have questions, please file a [support ticket](https://twilio.com/help/contact), or create a GitHub Issue in this repository.

### CODE REVIEW

File: src/main/java/com/twilio/exception/ApiException.java [hunk 1]
```diff
@@ -1,20 +1,23 @@
...
+import java.util.Map;
+
 public class ApiException extends TwilioException {
 
     private static final long serialVersionUID = -3228320166955630014L;
 
     private final Integer code;
     private final String moreInfo;
     private final Integer status;
+    private final Map<String, Object> details;
```
Comment (thinkingserious) [inline src/main/java/com/twilio/exception/ApiException.java:12]: Getting this warning from SonarLint: https://rules.sonarsource.com/java/RSPEC-1948

**Comment by ashish-s** [comment]
> When I run `make java` from within yoyodyne, I get many more changes not included in this PR. For example, [here](https://github.com/ashish-s/twilio-java/blob/error-details/src/main/java/com/twilio/rest/accounts/v1/credential/AwsCreator.java#L87), after generation, turns to:
> 
> ```java
> throw new ApiException(
>     restException.getMessage(),
>     restException.getCode(),
>     restException.getMoreInfo(),
>     restException.getStatus(),
>     restException.getDetails(),
>     null
> );
> ```
> 
> Could you please update [this file](https://github.com/ashish-s/twilio-java/blob/error-details/src/test/java/com/twilio/exception/ApiExceptionTest.java) also? Thanks!

running master branch of yoyodyne should not change anything. I have added tests for the new constructor in  ApiExceptionTest.java

File: src/main/java/com/twilio/exception/ApiException.java (hunk 1 above)
Comment (ashish-s) [reply src/main/java/com/twilio/exception/ApiException.java:12]: I can see the issue, why it is complaining, but the concrete type creation is in Jackson, and i wouldn't want to override that. Fwiw, jackson creates LinkedHashMap<> and its serializable

**Comment by thinkingserious** [comment]
The PR to `twilio-java` should be based on your [yoyodyne branch](https://github.com/ashish-s/twilio-java/tree/error-details).

**Comment by thinkingserious** [comment]
Your handcrafted code LGTM 👍 -- just need to get the generated code included. Thanks!

**Comment by ashish-s** [comment]
@thinkingserious thanks for the review, i added the generated code from my branch

File: src/test/java/com/twilio/exception/ApiExceptionTest.java [hunk 2]
```diff
@@ -29,13 +33,25 @@ public void twoArgConstructorShouldPreserveMessageAndCause() {
...
+
+    @Test
+    public void fullConstructorShouldPreserveAllValues() {
+        ApiException error = new ApiException(anyMessage, anyErrorCode, anyMoreInfo, anyHttpStatus, details, anyCause);
+        assertEquals("Message", anyMessage, error.getMessage());
+        assertSame("Cause", anyCause, error.getCause());
+        assertEquals("More info", anyMoreInfo, error.getMoreInfo());
+        assertEquals("Error code", anyErrorCode, error.getCode().intValue());
+        assertEquals("Status code", anyHttpStatus, error.getStatusCode().intValue());
+        assertEquals("Details", error.getDetails(), details);
```
Comment (thinkingserious) [inline src/test/java/com/twilio/exception/ApiExceptionTest.java:54]: nit: `assertEquals("Details", details, error.getDetails());`

File: src/main/java/com/twilio/rest/accounts/v1/credential/AwsCreator.java [hunk 3]
```diff
@@ -89,6 +89,7 @@ public Aws create(final TwilioRestClient client) {
                 restException.getCode(),
                 restException.getMoreInfo(),
                 restException.getStatus(),
+                restException.getDetails(),
```
Comment (childish-sambino) [inline src/main/java/com/twilio/rest/accounts/v1/credential/AwsCreator.java:92]: Can't we just pass `restException` into an `ApiException` constructor? Then future changes won't update so many files.

Comment (ashish-s) [reply src/main/java/com/twilio/rest/accounts/v1/credential/AwsCreator.java:92]: yeah thats a good idea, let me do that instead

File: src/test/java/com/twilio/exception/ApiExceptionTest.java (hunk 2 above)
Comment (ashish-s) [reply src/test/java/com/twilio/exception/ApiExceptionTest.java]: removed that constructor

File: src/main/java/com/twilio/rest/accounts/v1/credential/AwsCreator.java (hunk 3 above)
Comment (ashish-s) [reply src/main/java/com/twilio/rest/accounts/v1/credential/AwsCreator.java]: @childish-sambino removed the old constructor and added a new one with restException

File: src/main/java/com/twilio/exception/ApiException.java [hunk 4]
```diff
@@ -1,12 +1,15 @@
...
+import java.util.Map;
+
 public class ApiException extends TwilioException {
 
     private static final long serialVersionUID = -3228320166955630014L;
 
     private final Integer code;
     private final String moreInfo;
     private final Integer status;
+    private final Map<String,Object> details;
```
Comment (childish-sambino) [inline src/main/java/com/twilio/exception/ApiException.java:12]: nit: spacing `String, Object`

File: src/main/java/com/twilio/exception/RestException.java [hunk 5]
```diff
@@ -31,11 +33,13 @@
      */
     @JsonCreator
     private RestException(@JsonProperty("status") final int status, @JsonProperty("message") final String message,
-                          @JsonProperty("code") final Integer code, @JsonProperty("more_info") final String moreInfo) {
+                          @JsonProperty("code") final Integer code, @JsonProperty("more_info") final String moreInfo,
+                          @JsonProperty("details") final Map<String,Object> details) {
```
Comment (childish-sambino) [inline src/main/java/com/twilio/exception/RestException.java:37]: Same nit.

**Pull request merged**